File: iut/TP1_TP2/jour.py

# -*- coding: utf-8 -*-

def jour():

    jour = int(input("Entrez le jour : "))
    mois = int(input("Entrez le mois : "))
    annee = int(input("Entrez l'année : "))

    if mois < 1 or mois > 12:
        print("Mois invalide.")
        return

    if annee < 1600 or annee > 2199:
        print("Année hors limites.")
        return

    bissextile = (annee % 4 == 0 and annee % 100 != 0) or (annee % 400 == 0)

    jours_par_mois = [31, 29 if bissextile else 28, 31, 30, 31, 30,
                      31, 31, 30, 31, 30, 31]

    if jour < 1 or jour > jours_par_mois[mois - 1]:
        print("Jour invalide pour ce mois et cette année.")
        return

    deux_derniers = annee % 100

    somme = deux_derniers + (deux_derniers // 4)

    numeros_mois = [1, 4, 4, 0, 2, 5, 0, 3, 6, 1, 4, 6]
    somme += numeros_mois[mois - 1]

    if bissextile and (mois == 1 or mois == 2):
        somme -= 1


    if 1600 <= annee <= 1699:
        somme += 6
    elif 1700 <= annee <= 1799:
        somme += 4
    elif 1800 <= annee <= 1899:
        somme += 2
    elif 1900 <= annee <= 1999:
        somme += 0
    elif 2000 <= annee <= 2099:
        somme += 6
    elif 2100 <= annee <= 2199:
        somme += 4

    somme += jour
    reste = somme % 7


    jours_semaine = {
        1: "Dimanche",
        2: "Lundi",
        3: "Mardi",
        4: "Mercredi",
        5: "Jeudi",
        6: "Vendredi",
        0: "Samedi"
    }

    print(f"Le {jour:02d}/{mois:02d}/{annee} est un {jours_semaine[reste]}.")


jour()

File: iut/TP1_TP2/jour.c
#include <stdio.h>

int est_bissextile(int annee) {
    if ((annee % 4 == 0 && annee % 100 != 0) || (annee % 400 == 0)) {
        return 1;
    }
    return 0;
}

void jour() {
    
    int annee, mois, jour;

    printf("Entrez le jour: ");
    scanf("%d", &jour);
    printf("Entrez le mois: ");
    scanf("%d", &mois);   
    printf("Entrez l'année: ");
    scanf("%d", &annee);

    if (jour > 31 || (mois == 2 && jour >29))
    {
        printf("Erreur jours\n");
        return;
    }
    if (mois > 12)
    {
        printf("Erreur mois trop grand\n");
        return;
    }
    if (annee < 1600 || annee > 2100)
    {
        printf("Erreur mauvaise intervall\n");
        return;
    }
    

    int dernier_chiffre_annee = annee % 100;

    int quart = dernier_chiffre_annee / 4;
    int nombre_initial = dernier_chiffre_annee + quart;

    
    int mois_tableau[] = {1, 4, 4, 0, 2, 5, 0, 3, 6, 1, 4, 6};
    nombre_initial += mois_tableau[mois - 1];


    if (est_bissextile(annee) && (mois == 1 || mois == 2)) 
    {
        nombre_initial -= 1;
    }

    

    int valeur_siecle;
    if (annee >= 1600 && annee < 1700)
        valeur_siecle = 6;
    else if (annee >= 1700 && annee < 1800)
        valeur_siecle = 4;
    else if (annee >= 1800 && annee < 1900) 
        valeur_siecle = 2;
    else if (annee >= 1900 && annee < 2000)
        valeur_siecle = 0;
    else if (annee >= 2000 && annee < 2100) 
        valeur_siecle = 6;
    else if (annee >= 2100 && annee < 2200) 
        valeur_siecle = 4;
    

    nombre_initial += valeur_siecle;

    nombre_initial += jour;

    int jour_semaine = nombre_initial % 7;

    char* jours[] = {"Dimanche", "Lundi", "Mardi", "Mercredi", "Jeudi", "Vendredi", "Samedi"};
    
    printf("Le jour de la semaine est : %s\n", jours[jour_semaine]);
}

int main() {
    jour();
    return 0;
}